### ISSUE TITLE
Errors with NaN values

### ISSUE THREAD

**Issue by sandeshAdhikary** [jianda-chen/RAP_distance#2]
Firslty, thank you for providing the code for your method. I found the paper very interesting, and it is great to have a reference codebase to try out this method.

I am having trouble running experiments using the rap_distance. Namely, I keep getting NaN values - the policy is returning nan actions, which I assume is because of some gradient overflow issue during gradient updates.

I'm guessing this problem is coming from some numerical issue, maybe in the `_sqrt` or the `cosine_distance` function. I've played around with changing the EPSILON value to 1e-5 -- but am still running into NaNs.

Did you encounter these overflow issues as well? Any suggestions on preventing these NaN would be appreciated!

P.S.: I've tried to recreate the environment as closely as I could, except for the following changes:
1. I was running into errors installing the dm_control and dmc2gym dependencies in the conda_env.yaml. So I'm currently using the following
    - git+https://github.com/deepmind/dm_control.git
    - git+https://github.com/denisyarats/dmc2gym.git
2. I'm using a replay buffer of size 100_000 due to storage constraints.

**Comment by sandeshAdhikary** [comment]
Here's the bash script with arugments I'm using to run the experiment:

``
DOMAIN=cheetah
TASK=run
ACTION_REPEAT=4
SAVEDIR=../log

NOW=$(date +"%Y-%m-%d-%H-%M-%S")
echo ${SAVEDIR}/${DOMAIN}_${TASK}_${NOW}

python train.py \
    --domain_name ${DOMAIN} \
    --task_name ${TASK} \
    --agent 'RAP' \
    --init_steps 1000 \
    --replay_buffer_capacity 100000 \
    --num_train_steps 1000000 \
    --encoder_type pixel \
    --decoder_type identity \
    --transition_model_type 'probabilistic' \
    --action_repeat ${ACTION_REPEAT} \
    --critic_tau 0.01 \
    --encoder_tau 0.05 \
    --hidden_dim 1024 \
    --encoder_feature_dim 100 \
    --total_frames 1000 \
    --num_layers 4 \
    --num_filters 32 \
    --batch_size 128 \
    --encoder_lr 5e-4 \
    --decoder_lr 5e-4 \
    --actor_lr 5e-4 \
    --critic_lr 5e-4 \
    --alpha_lr 1e-4 \
    --alpha_beta 0.5 \
    --init_temperature 0.1 \
    --num_eval_episodes 5 \
    --work_dir ${SAVEDIR}/${DOMAIN}_${TASK}_${NOW} \
    --rap_structural_distance 'mico_angular' \
    --rap_reward_dist \
    --rap_square_target \
    --save_tb \
    --save_model \
    --seed 100 "$@"
    ``

**Comment by sandeshAdhikary** [comment]
I solved this issue by changing the `_sqrt` function. In the current implementation, the `tol` argument is taken as input but never used. Moreover, I think `tol` should be set to a small epsilon and not zero -- to avoid division by 0 during backprop.

Current implementation:
``
def _sqrt(x, tol=0.):
    tol = torch.zeros_like(x)
    return torch.sqrt(torch.maximum(x, tol))
``

Here's what I changed it to:
`def _sqrt(x, tol=EPSILON):
    tol = torch.ones_like(x)*tol
    return torch.sqrt(torch.maximum(x, tol))`

There may be other reasons why I was getting NaNs as well, but this seems to have solved it for now
